src/types/CanonicalField.ts

```diff
@@ -0,0 +1,57 @@
+import { ajv } from '../ajv';
+import type { JSONSchemaType } from 'ajv';
+
+export interface CanonicalField {
+  id: number;
+  label: string;
+  shortCode: string;
+  fieldType: string;
+  createdAt: Date;
+}
+
+export const defaultCanonicalField = {
```

Comment:
I've found it to be a good / useful habit of providing a "default" structure for various interfaces.  That way a dev who wants to create a canonical field can do something like:

```
import { defaultCanonicalField } from './types'

const myCanonicalField = {
  ...defaultCanonicalField,
  shortCode: 'whatever',
}
```

Reply:
This is probably the Java in me again, but if we go that route, why not provide a formal `class` with default values, constructors, etc.? That way whatever rules or conveniences we want to communicate, such as "specify this, not that", are all baked into that class?

Reply:
I hear ya.

Basically I think the root answer is "because of json".

More directly: generic object interface shape can be serialized and deserialized, and many of our tools are going to be working in terms of generic objects (e.g. express will provide our handlers generic objects parsed from json, the database will be providing generic objects parsed from sql results, etc.)

That means an Interface is not just more versatile, but is actually necessary regardless of whether or not we are going to also define a class.  Which is to say, we'd need the interface *regardless* in order to validate unknown data, but now we would also need to pass that type-validated data into a class constructor (and maintain a class in addition to the interface and JSONSchema).

More generally: TypeScript has moved away from classes and into more "functions with typed objects" and there are lots of other reasons for that (tree shaking is one -- if you provide an SDK and someone only needs the `createCanonicalField` method then a tree shaker can completely ignore all the other "canonical field" functions if those are defined as individual functions; if it's a class with a `delete` `update` etc method then the bundler can't selectively ditch those class methods just because they aren't used.

Reply:
I think that the right solution here is probably to only require things in the interface that are truly required, that way there's no need for a default object at all, since a default would never make sense (either the field needs to be set, or the field can be omitted).

I admit I am in the process of growing out of a bias against nulls (e.g. I have tended to prefer empty strings to nulls in almost all cases, since you can avoid having to worry about checking for null before use).

---------

Bottom line -- I think I'll take this out!

Reply:
Right, my suggestion was to keep the interface but then have a class implement that interface as a convenience rather than the default object as a convenience.

We are still left with trying to communicate the rules of optional/required/must-not-have via the interfaces, so it's probably six one way and half a dozen the other.

I had to look up tree shaking. At this moment my best summary is "it seems better to have relatively dumb data holders and independent functions that can operate on dumb data than to mash a bunch of data and behavior together into one all-or-nothing-package", e.g. a more functional style, less object-oriented?